File: classifier/data/raw/ai/samples/project_euler_zig/problem_041.zig
const std = @import("std");

// Project Euler 41: largest pandigital prime.
pub fn main() !void {
    var n: u32 = 7_654_321;
    while (n >= 1) : (n -= 2) {
        if (isPandigital7(n) and isPrime(n)) {
            printNumber(n);
            return;
        }
    }
}

fn isPandigital7(n: u32) bool {
    var value = n;
    var mask: u8 = 0;
    var count: u8 = 0;
    while (value != 0) : (value /= 10) {
        const digit: u8 = @intCast(value % 10);
        if (digit == 0 or digit > 7) return false;
        const bit: u8 = @as(u8, 1) << @intCast(digit - 1);
        if ((mask & bit) != 0) return false;
        mask |= bit;
        count += 1;
    }
    return count == 7 and mask == 0b111_1111;
}

fn isPrime(n: u32) bool {
    if (n < 2) return false;
    if (n % 2 == 0) return n == 2;
    var divisor: u32 = 3;
    while (divisor <= n / divisor) : (divisor += 2) {
        if (n % divisor == 0) return false;
    }
    return true;
}

fn printNumber(value: anytype) void {
    var buffer: [64]u8 = undefined;
    const text = std.fmt.bufPrint(&buffer, "{d}\n", .{value}) catch return;
    _ = std.c.write(1, text.ptr, text.len);
}

File: classifier/data/raw/ai/samples/project_euler_java/Problem041.java
// Project Euler Problem 41: largest pandigital prime.
public class Problem041 {
    public static void main(String[] args) {
        search(0, 0, 0);
    }

    private static boolean search(int position, int value, int used) {
        if (position == 7) {
            if (isPrime(value)) {
                System.out.println(value);
                return true;
            }
            return false;
        }
        for (int digit = 7; digit >= 1; digit--) {
            if ((used & (1 << digit)) == 0
                    && search(position + 1, value * 10 + digit, used | (1 << digit))) {
                return true;
            }
        }
        return false;
    }

    private static boolean isPrime(int n) {
        if (n < 2) return false;
        for (int d = 2; d * d <= n; d++) if (n % d == 0) return false;
        return true;
    }
}
